### PR TITLE
Solve chapter 2

### PR DESCRIPTION
**Pull request by fa7ad**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -348,9 +350,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
-
+firstHalf :: [a] -> [a]
+firstHalf = take =<< halfLen
```
Comment (vrom911) [inline src/Chapter2.hs:354]: Wow, very interesting solution!
Nice usage of the `Monad` instance for the arrow 🧠 🤯

File: src/Chapter2.hs [hunk 2]
```diff
@@ -348,9 +350,10 @@ Implement a function that returns only the first half of a given list.
...
+  where
+    halfLen = flip div 2 . length
```
Comment (vrom911) [inline src/Chapter2.hs]: You can use a function as an operator if you put it in the backticks. So you can avoid `flip`:

```haskell
`div` 2
```

File: src/Chapter2.hs [hunk 3]
```diff
@@ -727,7 +735,8 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate (x : xs) = replicate x x ++ smartReplicate xs
```
Comment (vrom911) [inline src/Chapter2.hs]: This is right! 💯 

You can notice that what is happening is you walk through each element of the list, change it to something else (by using `replicate` function) and then concatenate all of the results into one list.

In Haskell, there is the `map` function that does exactly the first part of it (walk through each element of the list, change it to something else ).

After using `map` you will get a list of lists. To flatten such a list, you can use the `concat` function:

```haskell
ghci> concat [[3, 1, 2], [4, 0, 1]]
[3,1,2,4,0,1]
```

This pattern of using `map` and `concat` is very common, so the standard library has even a standard function `concatMap` which is a combination of `map` and `concat` (as the name suggests).

File: src/Chapter2.hs [hunk 4]
```diff
@@ -780,13 +790,13 @@ Let's now try to eta-reduce some of the functions and ensure that we
 mastered the skill of eta-reducing.
 -}
 divideTenBy :: Int -> Int
-divideTenBy x = div 10 x
+divideTenBy = div 10
 
--- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Int -> [Int] -> [Int]
+listElementsLessThan = filter . (>)
```
Comment (vrom911) [inline src/Chapter2.hs:795]: You do love the point-free style 😄 👍🏼

File: src/Chapter2.hs [hunk 5]
```diff
@@ -841,7 +851,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n xs
+  | n < 0 = []
+  | otherwise = take (length xs) $ drop n $ cycle xs
```
Comment (vrom911) [inline src/Chapter2.hs]: Great!
If you want to speed up the solution, you can take `mod n (length xs)` 😉

File: src/Chapter2.hs [hunk 6]
```diff
@@ -857,7 +870,8 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind = foldl (\xs x -> x : xs) []
```
Comment (vrom911) [inline src/Chapter2.hs]: If you wish, you can use eta-reduction in this lambda as well 😸 
```suggestion
rewind = foldl (flip (:)) []
```

**Review by vrom911 — APPROVED** [review]
Awesome! Very interesting solutions 💠 

Feel free to merge this PR, it looks great 🚢

File: src/Chapter2.hs (hunk 2 above)
Comment (fa7ad) [reply src/Chapter2.hs]: I always forget about the infix notation lol, updated now 😄

File: src/Chapter2.hs (hunk 3 above)
Comment (fa7ad) [reply src/Chapter2.hs]: I didn't realize concatMap existed 😬. Updated the function to use concatMap now 😄

File: src/Chapter2.hs (hunk 4 above)
Comment (fa7ad) [reply src/Chapter2.hs:795]: Pointfree style is one of the few things that got me really interested in haskell tbh 😅

File: src/Chapter2.hs (hunk 1 above)
Comment (fa7ad) [reply src/Chapter2.hs:354]: Thanks 😄 
Had do some reading on the bind and reverse bind arrow to understand what it was doing, and then I was so happy that it worked 😆

**Comment by fa7ad** [comment]
@vrom911 Thanks for the review and informative suggestion. Learnt a bunch solving this chapter.
Merging now 😄

**Pull request merged**